### PR TITLE
NIFI-8499 - encrypted flow repository swap file implementation

### PR DESCRIPTION
**Pull request by greyp9**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Encrypted flow repository swap file implementation.  Provide ability to configure NiFi to use AES encryption when swapping flow files to disk.  

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+ * Test cases for {@link FileSystemSwapManagerEncrypt}.
+ */
+public class TestFileSystemSwapManagerEncrypt {
+    private static final Logger logger = Logger.getLogger(TestFileSystemSwapManagerEncrypt.class.getName());
+
+    /**
+     * Test a simple swap to disk / swap from disk operation.  Configured to use {@link StaticKeyProvider}.
+     */
+    @Test
+    public void testConfig_StaticKeyProvider_SerdeSuccess() throws GeneralSecurityException, EncryptionException, IOException {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java:62]: Recommend avoiding the `_` character in method names.  Perhaps something like `testSwapInSwapOut()`?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        // use temp folder on filesystem to temporarily hold swap content (clean up after test)
+        final File folderRepository = Files.createTempDirectory(getClass().getSimpleName()).toFile();
+        logger.info(folderRepository.getPath());
+        folderRepository.deleteOnExit();
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java:66]: Is this sufficient to delete the contents of the directory?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        properties.put(NiFiProperties.FLOWFILE_REPOSITORY_DIRECTORY, folderRepository.getPath());
+        final NiFiProperties nifiProperties = NiFiProperties.createBasicNiFiProperties(null, properties);
+
+        // generate some flow file content to swap to disk
+        final List<FlowFileRecord> flowFiles = new ArrayList<>();
+        for (int i = 0; (i < 10_000); ++i) {
+            flowFiles.add(new StandardFlowFileRecord.Builder().id(i).build());
+        }
+
+        // setup for test case
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java:83]: This comment and several others do not appear to add much in the way of clarification to the test behavior, should they be removed?

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the contribution @greyp9!  The core functionality looks clear and straightforward.  I included a number of detailed stylistic comments, but the overall approach looks good.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FileSystemSwapManagerEncrypt.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(FileSystemSwapManagerEncrypt.class);
+
+    private final SecretKey secretKey;
+
+    /**
+     * Default no args constructor for service loading only.
+     */
+    public FileSystemSwapManagerEncrypt() {
+        throw new IllegalStateException("ctor not supported; nifi properties not available");
+    }
```
[the remark this answers is not in the document]
Comment (greyp9) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FileSystemSwapManagerEncrypt.java:63]: I included the three constructors defined in the super class.  It seemed like a good idea to make the point explicitly that only one constructor is supported, as nifi properties are needed to resolve the repo encryption key.  But based on your suggestion, I've removed it.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FileSystemSwapManagerEncrypt.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    private static final Logger logger = LoggerFactory.getLogger(FileSystemSwapManagerEncrypt.class);
+
+    private final SecretKey secretKey;
+
+    /**
+     * Default no args constructor for service loading only.
+     */
+    public FileSystemSwapManagerEncrypt() {
+        throw new IllegalStateException("ctor not supported; nifi properties not available");
```
[the remark this answers is not in the document]
Comment (greyp9) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FileSystemSwapManagerEncrypt.java:62]: removed

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FileSystemSwapManagerEncrypt.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            cipher.init(Cipher.DECRYPT_MODE, secretKey, new GCMParameterSpec(SIZE_TAG_GCM_BITS, iv));
+            return new CipherInputStream(fis, cipher);
+        } catch (GeneralSecurityException e) {
+            throw new IOException(e);
+        }
+    }
+
+    protected OutputStream getOutputStream(final File file) throws IOException {
+        final byte[] iv = new byte[SIZE_IV_AES_BYTES];
+        new SecureRandom().nextBytes(iv);
```
[the remark this answers is not in the document]
Comment (greyp9) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FileSystemSwapManagerEncrypt.java:101]: Changed to use static instance.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FileSystemSwapManagerEncrypt.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        try {
+            final Cipher cipher = Cipher.getInstance(TRANSFORMATION);
+            cipher.init(Cipher.ENCRYPT_MODE, secretKey, new GCMParameterSpec(SIZE_TAG_GCM_BITS, iv));
+            return new CipherOutputStream(fos, cipher);
+        } catch (GeneralSecurityException e) {
+            throw new IOException(e);
+        }
+    }
+
+    private static final String ALGORITHM = "AES";
```
[the remark this answers is not in the document]
Comment (greyp9) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FileSystemSwapManagerEncrypt.java:113]: Leftover from working code.  Removed.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            }
+        });
+
+        return swapManager;
+    }
+
+    /**
+     * Borrowed from "nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManager.java".
+     */
+    public static class NopResourceClaimManager implements ResourceClaimManager {
```
[the remark this answers is not in the document]
Comment (greyp9) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java:133]: Copied from super class test.

[the remark this answers is not in the document]
Comment (greyp9) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java:133]: Replaced with Mockito usage.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        final Properties properties = new Properties();
+        properties.put(NiFiProperties.FLOWFILE_REPOSITORY_ENCRYPTION_KEY_PROVIDER_IMPLEMENTATION_CLASS, StaticKeyProvider.class.getName());
+        properties.put(NiFiProperties.FLOWFILE_REPOSITORY_ENCRYPTION_KEY_ID, NiFiProperties.FLOWFILE_REPOSITORY_ENCRYPTION_KEY);
+        properties.put(NiFiProperties.FLOWFILE_REPOSITORY_ENCRYPTION_KEY, StringUtils.repeat("00", 32));
+        properties.put(NiFiProperties.FLOWFILE_REPOSITORY_DIRECTORY, folderRepository.getPath());
+        final NiFiProperties nifiProperties = NiFiProperties.createBasicNiFiProperties(null, properties);
+
+        // generate some flow file content to swap to disk
+        final List<FlowFileRecord> flowFiles = new ArrayList<>();
+        for (int i = 0; (i < 10_000); ++i) {
```
[the remark this answers is not in the document]
Comment (greyp9) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java:79]: Copied the number from the test of the super class.  Replaced with 100.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java (hunk 2 above)
Comment (greyp9) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/TestFileSystemSwapManagerEncrypt.java:66]: In my testing, yes.  The "swap in" operation removes the swap file.  And the "deleteOnExit()" documentation suggests this ordering of the temp folder and the "swap" folder contained within.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/EncryptedFileSystemSwapManager.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+import java.security.SecureRandom;
+
+/**
+ * <p>
+ * An implementation of {@link FlowFileSwapManager} that swaps FlowFiles
+ * to/from local disk.  The swap file is encrypted using AES/GCM, using the
+ * encryption key defined in nifi.properties for the FlowFile repository.
+ * </p>
+ */
+@SuppressWarnings("unused")  // instantiation via reflection in product
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/EncryptedFileSystemSwapManager.java:50]: Is this annotation still necessary now that the other constructors are not included?

**Review by exceptionfactory — COMMENTED** [review]
Thanks making the updates @greyp9!  I noted one additional question on an annotation, but otherwise it looks good.  I will go through some additional runtime verification, but it looks like it is close to completion.

**Review by exceptionfactory — APPROVED** [review]
Thanks for making the additional changes @greyp9!  Tested runtime configuration and confirmed functionality. +1 Merging.

**Pull request closed without merging**